### ISSUE TITLE
Feature / Help Request: Implementing Constraint Decoding in AnemLL

### ISSUE THREAD

**Issue by ShubhamMishra1611** [Anemll/Anemll#37]
Hi team 👋

I’ve been experimenting with **AnemLL** for a project that requires **structured generation** — mainly ensuring valid JSON output during inference (for iOS integration).

To validate the approach, I first implemented the logic in Python using the Transformers library. The Python version works well — it tracks field states, applies logit masking, and maintains proper JSON formatting even with smaller models like **Gemma-270M**.

Now, I’m trying to port the same logic into **AnemLL (Swift)**. However, the model output becomes gibberish, e.g.

```json
{"name": ";;;;;;;", "tags": ";;;;;", "age": 26}
```

---

### ⚙️ Changes Made to AnemLL Core

To implement constraint decoding, I modified the core logic as follows:

1. **Added** `generateNextTokenWithConstraints`

   * Similar to `generateNextToken`, but applies **logit masking** before sampling the next token.
   * Used for enforcing token-level constraints during generation.

2. **Added** `getLogitsForTokens`

   * Utility to **expose logits** for given tokens.
   * Helps externally manage or inspect token probabilities during guided generation.

3. **Integrated new helper files**

   * These use the above two functions to perform the **constraint logic** and state tracking.
   * All JSON-related rules and validation happen here.

---

### 🧠 Current Issue

Even though the logic seems correct (matching the working Python version), the generated output from the model becomes incoherent.
I’m unsure whether:

* The **constraint decoding logic** is disrupting the inference flow, or
* The issue lies in how logits are being retrieved/manipulated in the Swift layer.

---

### 🔍 What I’m Looking For

I’d really appreciate some guidance or feedback on:

* Whether this approach aligns with how AnemLL handles token generation internally
* Possible areas in the inference pipeline where logit manipulation could break behavior
* Any recommended way to integrate custom decoding logic into AnemLL

Once this is stable, I’d love to open a PR to contribute the feature — it could help others implementing structured / JSON-based decoding.

---

### 📎 References & Attachments

* My fork with the core changes: [[ShubhamMishra1611/Anemll](https://github.com/ShubhamMishra1611/Anemll)](https://github.com/ShubhamMishra1611/Anemll)
* Python reference logic: 

[constraint_generation.py](https://github.com/user-attachments/files/22880139/constraint_generation.py)

---

Thanks for maintaining this awesome project 🙏
**— Shubham Mishra**

**Comment by hermgir1bot** [comment]
Thanks Shubham — constraint/structured decoding is a great fit and we'd welcome the PR. On where the gibberish likely comes from, the Swift reference exposes token generation in `anemll-swift-cli/Sources/AnemllCore/InferenceManager.swift`:

- `generateNextToken` (line ~2234) is where logits come back and sampling happens — this is the path that supports logit masking.
- `generateNextTokenArgmaxSync` (line ~3028) is the **in-model-argmax** path that returns only the winner index (no full logits). If your model has `argmax_in_model: true`, you **cannot** do logit masking there — you'd need the non-argmax path.

Also check that your masking runs *before* the temperature/top-p step, and that you advance the KV-cache state by exactly one position per token — if state desyncs you get exactly the kind of repetitive/repeating output you saw.

Keeping this open as an `enhancement`; your fork + `constraint_generation.py` is a solid reference to diff against.
